java-checks/src/main/java/org/sonar/java/checks/JacksonDeserializationCheck.java

```diff
@@ -0,0 +1,124 @@
+public class JacksonDeserializationCheck extends IssuableSubscriptionVisitor {
+
+  private static final MethodMatcherCollection ENABLE_DEFAULT_TYPING = MethodMatcherCollection.create(
+      MethodMatcher.create()
+          .typeDefinition("com.fasterxml.jackson.databind.ObjectMapper")
+          .name("enableDefaultTyping")
+          .withoutParameter(),
+      MethodMatcher.create()
+          .typeDefinition("org.codehaus.jackson.map.ObjectMapper")
+          .name("enableDefaultTyping")
+          .withoutParameter()
+      );
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD_INVOCATION, Tree.Kind.ANNOTATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(Tree.Kind.METHOD_INVOCATION) && ENABLE_DEFAULT_TYPING.anyMatch((MethodInvocationTree) tree)) {
+      reportIssue(tree, "Consider using @JsonTypeInfo instead of enabling polymorphic type handling globally");
+    } else if (tree.is(Tree.Kind.ANNOTATION)) {
+      AnnotationTree annotationTree = (AnnotationTree) tree;
+      if (isJsonTypeInfo(annotationTree) && isAnnotationOnClassOrField(annotationTree)) {
+        findUseArgument(annotationTree).ifPresent(useAnnotationArgument ->
+            reportIssue(useAnnotationArgument, "Use @JsonTypeInfo(use = Id.NAME) instead"));
+      }
+    }
+  }
+
+  private static boolean isJsonTypeInfo(AnnotationTree annotationTree) {
+    Type annotationType = annotationTree.annotationType().symbolType();
+    return annotationType.is("com.fasterxml.jackson.annotation.JsonTypeInfo")
+        || annotationType.is("org.codehaus.jackson.annotate.JsonTypeInfo");
+  }
+
+  private static boolean isAnnotationOnClassOrField(AnnotationTree annotationTree) {
+    if (annotationTree.parent().is(Tree.Kind.MODIFIERS)) {
+      Tree modifiers = annotationTree.parent();
+      return modifiers.parent().is(Tree.Kind.CLASS, Tree.Kind.VARIABLE);
+    }
+    return false;
+  }
+
+  private static Optional<ExpressionTree> findUseArgument(AnnotationTree annotationTree) {
+    for (ExpressionTree tree : annotationTree.arguments()) {
+      if (tree.is(Tree.Kind.ASSIGNMENT)) {
+        AssignmentExpressionTree assignment = (AssignmentExpressionTree) tree;
+        if (((IdentifierTree) assignment.variable()).name().equals("use")
+            && isJsonTypeIdEnumValue(assignment.expression(),"CLASS", "MINIMAL_CLASS")) {
+          return Optional.of(assignment.expression());
+        }
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static boolean isJsonTypeIdEnumValue(ExpressionTree tree, String... value) {
+    if (!isJsonTypeId(tree)) {
+      return false;
+    }
+    String valueName;
+    if (tree.is(Tree.Kind.MEMBER_SELECT)) {
+      valueName = ((MemberSelectExpressionTree) tree).identifier().name();
+    } else {
+      valueName = ((IdentifierTree) tree).name();
+    }
+    for (String enumValue : value) {
```

Comment:
nitpick: if you pass `value` as a Set, you can just `return VALUE_SET.contains(valueName);` instead of for + if + two return statements

Reply:
nice suggestion, thanks!